320add increment absolute

PuzzledFace opened the pull request:
closes #320

wwojciech commented on R/Rules-class.R:
```diff
@@ -695,6 +695,108 @@ setClass(
+#' @export
+.IncrementsAbsolute <- setClass(
+    Class="IncrementsAbsolute",
+    representation(
```
better to use `slots = c(....)` instead of `representation`

wwojciech commented on R/Rules-class.R:
```diff
@@ -695,6 +695,108 @@ setClass(
+      intervals="numeric",
+      increments="numeric"
+    ),
+    prototype(
```
lets make it explicitly `prototype = prototype(....)`

wwojciech commented on R/Rules-class.R:
```diff
@@ -695,6 +695,108 @@ setClass(
+      increments=c(2, 1)
+    ),
+    contains="Increments"
+)
```
can you pl. add validation function for this class?

wwojciech commented on R/Rules-class.R:
```diff
@@ -695,6 +695,108 @@ setClass(
+                      ) {
+  .IncrementsAbsolute(intervals=intervals,
+                      increments=increments)
+}# IncrementsAbsoluteDLT ----
```
Please use `styler` package to better format this code.
use [`CLASS_NAME`] instead of `code{\....}`

wwojciech commented on R/Rules-class.R:
```diff
@@ -695,6 +695,108 @@ setClass(
+                         ) {
+  .IncrementsAbsoluteDLT(intervals=intervals,
+                         increments=increments)
+}
```
same comments apply to this class/constructors.

wwojciech commented on R/Rules-methods.R:
```diff
@@ -1033,6 +1033,81 @@ setGeneric("maxDose",
+      # print(paste0("Empty grid: returning dose index ", increments@increments[1], ": ", data@doseGrid[increments@increments[1]]))
+      return(data@doseGrid[min(length(data@doseGrid), increments@increments[1])])
+    }
+    usedMax <- max(data@x)
```
we use snake style rather than camel for naming variables, so it should be `used_max`.

wwojciech commented on R/Rules-methods.R:
```diff
@@ -1033,6 +1033,81 @@ setGeneric("maxDose",
+    # print(paste0("usedMax: ", usedMax))
+    # print(paste0("usedMaxIndex: ", usedMaxIndex))
+    # print(paste0("Current interval: ", currentInterval))
+    allowedIndex <- min(length(data@doseGrid), usedMaxIndex + increments@increments[currentInterval])
```
lets remove any code that is commented.

wwojciech commented on R/Rules-validity.R:
```diff
@@ -224,6 +224,54 @@ v_next_best_max_gain_samples <- function(object) {
+  v$result()
+}
+#' Internal Helper Functions for Validation of [`Increments`] Objects
+#'
```
why do we need this?

PuzzledFace replied:
Typo.  Removed.

wwojciech commented on examples/Rules-method-maxDose-IncrementsAbsolute.R:
```diff
@@ -0,0 +1,17 @@
+
+# Create the data
+data <- Data(x=c(0.1, 0.5, 1.5, 3, 6, 8, 8, 8),
```
please style the code with styler, and check it with linter.

PuzzledFace replied:
styled the new code with `stylr`.

`lintr` reported an error regarding a missing `nolint start`.  I added one at line 983.

`lintr` then reported  several "variable and function name style should be snake_case or symbols" warnings for the new code.  This is consistent with the same warnings reported elsewhere.  I assume this is acceptable.

wwojciech reviewed (changes requested):
Thank you! Can you also write unit tests for the classes?

PuzzledFace commented:
> Thank you! Can you also write unit tests for the classes?

I added tests to test-Rules-class.R an test-Rules-methods.R.  I'm not sure why they are not showing up here.

wwojciech commented:
> > Thank you! Can you also write unit tests for the classes?
> 
> I added tests to test-Rules-class.R an test-Rules-methods.R. I'm not sure why they are not showing up here.

I checked this file in this branch, and there is not update for these files, so it must be that the changes were not stored/transffered in/to this branch.

wwojciech commented on R/Rules-class.R:
```diff
@@ -718,34 +718,36 @@ setClass(
 #' @param intervals (`numeric`) \cr a vector of right hand boundaries of the intervals
 #' @param increments (`numeric`) \cr a vector of the maximum increment allowed in the corresponding interval
-#' @return the \code{\linkS4class{IncrementsAbsolute}} object
+#' @return the [\linkS4class{IncrementsAbsolute}] object
```
```suggestion
#' @return the [`IncrementsAbsolute`] object
```

PuzzledFace commented:
> > > Thank you! Can you also write unit tests for the classes?
> > 
> > 
> > I added tests to test-Rules-class.R an test-Rules-methods.R. I'm not sure why they are not showing up here.
> 
> I checked this file in this branch, and there is not update for these files, so it must be that the changes were not stored/transferred in/to this branch.

I'm not sure what happened with the updated test files.  The changes were there in my local repo, but the files were not flagged as uncommitted, nor were the changes on the remote repo.  All should be good now.

I believe I have now  addressed all review comments.

wwojciech commented:
> > > > Thank you! Can you also write unit tests for the classes?
> > > 
> > > 
> > > I added tests to test-Rules-class.R an test-Rules-methods.R. I'm not sure why they are not showing up here.
> > 
> > 
> > I checked this file in this branch, and there is not update for these files, so it must be that the changes were not stored/transferred in/to this branch.
> 
> I'm not sure what happened with the updated test files. The changes were there in my local repo, but the files were not flagged as uncommitted, nor were the changes on the remote repo. All should be good now.
> 
> I believe I have now addressed all review comments.

Hi John, thanks a lot. However, I see that some of the review comments are still not addressed. For instance, the one about missing validation function for classes.

danielinteractive commented:
Hi guys it might be most efficient when you have a quick 1:1 meeting to resolve the last bits and pieces together online

PuzzledFace commented:
> Hi guys it might be most efficient when you have a quick 1:1 meeting to resolve the last bits and pieces together online

The issue was the "missing" validation code was the same as before: the files were changed in my local copy of the repo but were not staged, committed for pushed to the remote repo for some reason.

I'm happy to talk with Wojciech one-to-one if that helps.

wwojciech commented:
> > Hi guys it might be most efficient when you have a quick 1:1 meeting to resolve the last bits and pieces together online
> 
> The issue was the "missing" validation code was the same as before: the files were changed in my local copy of the repo but were not staged, committed for pushed to the remote repo for some reason.
> 
> I'm happy to talk with Wojciech one-to-one if that helps.

Hi @PuzzledFace, @danielinteractive 
Thanks, apart from some minor issues (that I can correct by myself), the biggest is that the validation for classes is still missing and also the methods are not clear enough for me (as I pointed out in my review).
Let's have some 1:1 chat today, I will contact you on Slack to agree on details. Tx!

wwojciech commented:
> > > Hi guys it might be most efficient when you have a quick 1:1 meeting to resolve the last bits and pieces together online
> > 
> > 
> > The issue was the "missing" validation code was the same as before: the files were changed in my local copy of the repo but were not staged, committed for pushed to the remote repo for some reason.
> > I'm happy to talk with Wojciech one-to-one if that helps.
> 
> Hi @PuzzledFace, @danielinteractive Thanks, apart from some minor issues (that I can correct by myself), the biggest is that the validation for classes is still missing and also the methods are not clear enough for me (as I pointed out in my review). Let's have some 1:1 chat today, I will contact you on Slack to agree on details. Tx!

Hey @PuzzledFace Thank you very the great discussion we had. Please let me know when you think you are ready with the changes you are about to apply. Tx!

wwojciech commented:
Hi @PuzzledFace - I have just merged the latest `main` branch into your branch, as we updated the `main` significantly in view of recent changes to the `lintr` package.

danielinteractive commented:
@PuzzledFace @wwojciech what is the status here? let's try to finish this, before starting many other new things

wwojciech commented:
> @PuzzledFace @wwojciech what is the status here? let's try to finish this, before starting many other new things

Hi @danielinteractive The status is that is is after review and waiting for review comments to be implemented.

danielinteractive reviewed (commented):
almost there :-)

wwojciech commented:
> @wwojciech can you please re-review and approve if all good?

Sure @danielinteractive - but just waiting for a few changes to be implemented here (according to my PR). One these are ready - I will review it.

danielinteractive commented:
Hi @PuzzledFace did you have the chance already to do a new branch/PR?

PuzzledFace commented:
> Hi @PuzzledFace did you have the chance already to do a new branch/PR?

Not yet.  Is the delay causing you problems?

danielinteractive commented:
No, it does not block anything. Just trying to clean up the board slowly :-)
